stan/math/prim/functor/hcubature.hpp

```diff
@@ -0,0 +1,521 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files
+// (the "Software"), to deal in the Software without restriction,
+// including without limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of the Software,
+// and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+
+#ifndef STAN_MATH_PRIM_FUNCTOR_HCUBATURE_HPP
+#define STAN_MATH_PRIM_FUNCTOR_HCUBATURE_HPP
+
+#include <set>
+#include <stan/math/prim/fun/choose.hpp>
+#include <stan/math/prim/fun/isfinite.hpp>
+#include <stan/math/prim/fun/fabs.hpp>
+#include <stan/math/prim/fun/fmax.hpp>
+#include <stan/math/prim/fun/max.hpp>
+
+namespace stan {
+namespace math {
+
+namespace internal {
+
+// tools
+static const double xd7[8] = {-9.9145537112081263920685469752598e-01,
+                              -9.4910791234275852452618968404809e-01,
+                              -8.6486442335976907278971278864098e-01,
+                              -7.415311855993944398638647732811e-01,
+                              -5.8608723546769113029414483825842e-01,
+                              -4.0584515137739716690660641207707e-01,
+                              -2.0778495500789846760068940377309e-01,
+                              0.0};
+
+static const double wd7[8] = {2.2935322010529224963732008059913e-02,
+                              6.3092092629978553290700663189093e-02,
+                              1.0479001032225018383987632254189e-01,
+                              1.4065325971552591874518959051021e-01,
+                              1.6900472663926790282658342659795e-01,
+                              1.9035057806478540991325640242055e-01,
+                              2.0443294007529889241416199923466e-01,
+                              2.0948214108472782801299917489173e-01};
+
+static const double gwd7[4] = {1.2948496616886969327061143267787e-01,
+                               2.797053914892766679014677714229e-01,
+                               3.8183005050511894495036977548818e-01,
+                               4.1795918367346938775510204081658e-01};
+
+struct one_d {
+  double result;
+  double err;
+  int kdivide = 0;
+};
+
+struct GenzMalik {
+  std::vector<double*> p[4];
+  double w[5];
+  double wd[4];
+};
+
+template <typename T_c, typename T_n, typename T_p, typename T_x>
+void combination(T_c& c, const T_n& n, const T_p& p, const T_x& x) {
+  size_t i, r, k = 0;
+  for (i = 0; i < p - 1; i++) {
+    c[i] = (i != 0) ? c[i - 1] : 0;
+    do {
+      c[i]++;
+      r = choose(n - c[i], p - (i + 1));
+      k = k + r;
+    } while (k < x);
+    k = k - r;
+  }
+  if (p > 1)
+    c[p - 1] = c[p - 2] + x - k;
+  else
+    c[0] = x;
```

Comment:
We generally follow the [google style guide](https://google.github.io/styleguide/cppguide.html) so curly braces for if/elses, while, and for loops

```suggestion
  if (p > 1) {
    c[p - 1] = c[p - 2] + x - k;
  } else {
    c[0] = x;
  }
```

clang-format has an option to do that for you if you don't want to do it manually

Reply:
Comes in next commit.